Feature/redistribution

odlomax opened the pull request:
This PR extends the Redistribution class functionality to NodeColumns, CellColumns, EdgeColumns and PointCloud, in addition to StructuredColumns. It now also gets the field shape from the field object, instead of the functionspace object. This means that you don't need multiple redistribution objects for different numbers of levels. It can currently redistribute fields of rank 1, rank 2 and rank 3 (i.e., up to 3D vector fields).

The redistribution factory defaults to this new implementation. If you want to use the old version, you can do so by supplying a config to the constructor:
```
 auto redist = atlas::Redistribution( sourceFunctionSpace, targetFunctionSpace,
                                         util::Config( "type", "RedistributeStructuredColumns" ) )
```
However, given the limitations of the old implementation, it's probably not worth it. I suspect there's little performance benefit for non-enormous fields.

The implementation works by gathering the global indices across PEs, then creating the intersections of the source global indices on the local PE and the target global indices on all the other PEs (and vice versa). Fields can then be redistributed via an MPI allToAllv. The setup method inspects the global indices of all nodes/elements which have ghost = 0. If a global index field is missing (i.e., for PointClouds) a unique ID is constructed from the lonlat field. If a ghost field is missing, then one is created via `partition != thisPartition || remoteIdx != localIdx`.

Currently the implementation cannot redistribute a CellColumns functionspace for a StructuredGrid as there is no  `build_cells_global_idx` method implemented in `mesh/actions/BuildParallelFields.cc`. It can, however, redistribute an EdgeColumns functionspace for the same mesh, if halo = 0.

Summary of what _does_ work:

* StructuredGrid - NodeColumns
* StructuredGrid - EdgeColumns ( halo = 0 only )
* StructuredGrid - StructuredColumns
* PointCloud (lonlats harvested from another functionspace)
* CubedSphere - NodeColumns
* CubedSphere - CellColumns

Here is an example of a redistribution between a CubedSphere NodeColumns functionspace with an "equal_regions" partitioner, and the same with a "cubedsphere" partitioner:

Source:
![source](https://user-images.githubusercontent.com/20695114/136710984-1acef96c-8361-4455-8526-496efa7c1f8f.png)

Target:
![target](https://user-images.githubusercontent.com/20695114/136710990-c8ed3083-96cb-46eb-b643-3e0d5f4880e4.png)

Closes #69

MarekWlasak committed:
[Github #76] Fix TransIFS packing of StructuredColumns fields with halos

wdeconinck committed:
Fix compiler issue with Ubuntu1604 and GNU 5.4

odlomax committed:
Refactored redistribution factory.

wdeconinck committed:
Fix compiler issue with Ubuntu1604 and GNU 5.4

wdeconinck committed:
Fix compiler issue with Ubuntu1604 and GNU 5.4

wdeconinck committed:
Github Actions caching with intel depends on version of intel compiler

wdeconinck committed:
Remove deepcode badge

odlomax committed:
ATLAS-332 [Github #77] Implement CubedSphere function spaces

odlomax committed:
saving before merge

odlomax committed:
saving before merge

odlomax committed:
Merge branch 'feature/cubed-sphere-base2' into feature/redistribution

odlomax committed:
Ready for PR

odlomax committed:
Added tests for int and long types.

odlomax committed:
saving before merge.

odlomax committed:
Merge branch 'feature/cubed-sphere-base2' into feature/redistribution

odlomax committed:
Added uniqueness debug check.

odlomax committed:
Fixed CubedSphere CellColumns test.

odlomax committed:
Corrected functionspace SFINAE rules.

odlomax committed:
Corrected comment.

wdeconinck committed:
Fix clang-tidy warnings

wdeconinck committed:
ATLAS-333 Add remote_index and global_index access to FunctionSpace

pmaciel committed:
Fix warnings

odlomax committed:
Tidy up.

MarekWlasak commented on src/atlas/redistribution/detail/RedistributeGeneric.cc:
```diff
@@ -0,0 +1,498 @@
+Field getUidField( const FunctionSpaceType* functionSpaceImpl ) {
+
+    // Make UID field.
+    Field uid = Field( " Uniquie ID ", array::make_datatype<uidx_t>(),
```
Unique ID

odlomax replied:
Done.

MarekWlasak commented on src/atlas/redistribution/detail/RedistributeGeneric.cc:
```diff
@@ -0,0 +1,498 @@
+#include "atlas/functionspace/EdgeColumns.h"
+#include "atlas/functionspace/NodeColumns.h"
+#include "atlas/functionspace/PointCloud.h"
+#include "atlas/functionspace/StructuredColumns.h"
```
Why is CubedSphereFunctionSpace not in the list?

odlomax replied:
CubeSphereNodeColumns and CubedSphereCellColumns are specialisations of NodeColumns and CellColumns, so they're covered by these headers.

MarekWlasak commented on src/atlas/redistribution/detail/RedistributeGeneric.cc:
```diff
@@ -0,0 +1,498 @@
+    const auto ghostView = array::make_view<int, 1>( ghost );
+    const auto uidView = array::make_view<uidx_t, 1>( uid );
+
+    auto uidVec = std::vector<IdxUid>{};
```
should you be reserving memory here?

wdeconinck replied:
Agreed. You should definitely reserve here with `functionSpaceImpl->size()`

odlomax replied:
Good point. I've now called reserve on vectors wherever I can.

MarekWlasak commented on src/atlas/redistribution/detail/RedistributeGeneric.cc:
```diff
@@ -0,0 +1,498 @@
+            continue;
+        }
+       uidVec.push_back( IdxUid( i, uidView( i ) ) );
+    }
```
Is there a better way than the above?  It feels quite Fortranesque!

wdeconinck replied:
And this could be instead:
```
if( not ghostView(i) ) {
    uidVec.emplace_back( i, uidView(i) );
}
```
`emplace_back` has variadic arguments that forwards to the constructor and does an in-place construction, rather than the original copy (or perhaps move) construction

odlomax replied:
I agree. `emplace_back` is the better option here.

odlomax committed:
Aggressive memory reservation for vectors.

odlomax committed:
Corrected typo.

benjaminmenetrier committed:
Fortran interfaces for Projection methods (#79)

wdeconinck committed:
Cleanup RedistributeStructuredColumns

Do not use the implementations of FunctionSpace, but rather use the Handle
This avoids explicit casting

wdeconinck committed:
Merge branch 'develop' into feature/redistribution

wdeconinck committed:
Remove explicit casting and needless sfinae from RedistributeGeneric

wdeconinck commented:
I consider this ready for upstream PR now

wdeconinck committed:
Set timeout of atlas_test_cubedsphere_meshgen to 30s

odlomax commented on src/atlas/redistribution/detail/RedistributeGeneric.cc:
```diff
@@ -0,0 +1,447 @@
+
+    // Note there is a difference between uid_field and global_index in the type: uidx_t vs gidx_t
+    // Currently this aliases the same type though...
+
```
I spotted that also. Apologies for living dangerously!

odlomax commented on src/atlas/redistribution/detail/RedistributeGeneric.cc:
```diff
@@ -0,0 +1,447 @@
+    }
+    if( functionspace::EdgeColumns( functionspace ) ||
+        functionspace::CellColumns( functionspace ) ) {
+
```
Good old SFINAE. Clever, but often pointless!

wdeconinck replied:
It has its purpose, but was probably overkill here, and we can work with the handles instead.

odlomax replied:
Oh I agree. And what you did fits much better with other polymorphism patterns in Atlas.

Still, I enjoyed my moment as a mad sfinae scientist! 😂

wdeconinck commented:
Merged upstream
